File: src/ABTD/actuator_gains/pid.cpp
#include <iostream>

class PIDController {
public:
  // Constructor
  PIDController(double Kp, double Ki, double Kd) :
      Kp_(Kp), Ki_(Ki), Kd_(Kd), integral_(0.0), previous_error_(0.0) {}

  // Update function to calculate control output
  double update(double current_value, double target_value, double dt) {
    // Calculate error
    double error = target_value - current_value;

    // Update integral term (avoid integral wind-up)
    integral_ += error * dt;
    if (integral_ > integral_limit_) {
      integral_ = integral_limit_;
    } else if (integral_ < -integral_limit_) {
      integral_ = -integral_limit_;
    }

    // Calculate derivative term
    double derivative = (error - previous_error_) / dt;
    if (derivative > derivative_limit_) {
      derivative = derivative_limit_;
    } else if (derivative < -derivative_limit_) {
      derivative = -derivative_limit_;
    }

    // Update previous error
    previous_error_ = error;

    // Calculate control output
    double control_output = Kp_ * error + Ki_ * integral_ + Kd_ * derivative;
    if (control_output > output_limit_) {
      control_output = output_limit_;
    } else if (control_output < -output_limit_) {
      control_output = -output_limit_;
    }

    return control_output;
  }

private:
  // PID gains
  double Kp_, Ki_, Kd_;

  // Integral term
  double integral_;

  // Previous error
  double previous_error_;

  // Optional limits for integral and output
  double integral_limit_ = 1.0;
  double derivative_limit_ = 1.0;
  double output_limit_ = 2.0;
};

int main() {
  // Sample usage
  PIDController controller(0.1, 0.5, 0.3); // Adjust gains as needed

  double current_value = 0.0;
  double target_value = 10.0;
  double dt = 0.1;

  for (int i = 0; i < 10^2; ++i) {
    double control_output = controller.update(current_value, target_value, dt);
    current_value += control_output*dt;

    std::cout << "target_value: " << target_value << ", current_value: " << current_value << ", control_output: " << control_output << std::endl;

    target_value += 2.0/(i+1);
  }

  return 0;
}

File: src/ABTD/actuator_gains/pid.py
class PIDController:
   def __init__(self, Kp, Ki, Kd):
       self.Kp = Kp
       self.Ki = Ki
       self.Kd = Kd
       self.integral = 0.0
       self.previous_error = 0.0
       self.integral_limit = 1.0
       self.derivative_limit = 1.0
       self.output_limit = 2.0

   def update(self, current_value, target_value, dt):
       error = target_value - current_value

       # Update integral term with anti-windup
       self.integral += error * dt
       self.integral = max(min(self.integral, self.integral_limit), -self.integral_limit)

       # Calculate derivative term with clamping
       derivative = (error - self.previous_error) / dt
       derivative = max(min(derivative, self.derivative_limit), -self.derivative_limit)

       self.previous_error = error

       # Calculate control output with clamping
       control_output = self.Kp * error + self.Ki * self.integral + self.Kd * derivative
       control_output = max(min(control_output, self.output_limit), -self.output_limit)

       return control_output

if __name__ == "__main__":
   controller = PIDController(0.1, 0.5, 0.3)  # Adjust gains as needed

   current_value = 0.0
   target_value = 10.0
   dt = 0.1

   for i in range(1000):  # Pythonic for loop
       control_output = controller.update(current_value, target_value, dt)
       current_value += control_output * dt

       print(f"target_value: {target_value:.2f}, current_value: {current_value:.2f}, control_output: {control_output:.2f}")

       target_value += 2.0 / (i + 1)
